### PR TITLE
KAFKA-15883: Implement RemoteCopyLagBytes

### PR DESCRIPTION
**Pull request by clolov**
### Summary
This pull request implements the first in the list of metrics in [KIP-963: Additional metrics in Tiered Storage](https://cwiki.apache.org/confluence/display/KAFKA/KIP-963%3A+Additional+metrics+in+Tiered+Storage).

Since each partition of a topic will be serviced by its own RLMTask we need an aggregator object for a topic. The aggregator object in this pull request is `BrokerTopicAggregatedMetric`. Since the RemoteCopyLagBytes is a gauge I have introduced a new `GaugeWrapper`. The GaugeWrapper is used by the metrics collection system to interact with the BrokerTopicAggregatedMetric. The RemoteLogManager interacts with the BrokerTopicAggregatedMetric directly.

### CODE REVIEW

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 1]
```diff
@@ -283,8 +283,39 @@ class BrokerTopicMetrics(name: Option[String], configOpt: java.util.Optional[Kaf
       meter()
   }
 
+  case class GaugeWrapper(metricType: String, brokerTopicAggregatedMetric: BrokerTopicAggregatedMetric) {
+    @volatile private var lazyGauge: Gauge[Long] = _
+    private val gaugeLock = new Object
+
+    def gauge(): Gauge[Long] = {
```
Comment (showuon) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: When do we initialize the gauge for topic with tags?

**Review by showuon — COMMENTED** [review]
@clolov , thanks for the PR. This implementation makes sense to me. Could we also add integration test for it?

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 1 above)
Comment (clolov) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Good question, I believe this is a leftover from the meters. I have removed it. I have further added to an integration test. To be honest, however, I haven't found a nice way to simulate an upload to remote storage being blocked to showcase the increase in this value via a test.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 2]
```diff
@@ -299,7 +329,7 @@ class BrokerTopicMetrics(name: Option[String], configOpt: java.util.Optional[Kaf
     BrokerTopicStats.NoKeyCompactedTopicRecordsPerSec -> MeterWrapper(BrokerTopicStats.NoKeyCompactedTopicRecordsPerSec, "requests"),
     BrokerTopicStats.InvalidMagicNumberRecordsPerSec -> MeterWrapper(BrokerTopicStats.InvalidMagicNumberRecordsPerSec, "requests"),
     BrokerTopicStats.InvalidMessageCrcRecordsPerSec -> MeterWrapper(BrokerTopicStats.InvalidMessageCrcRecordsPerSec, "requests"),
-    BrokerTopicStats.InvalidOffsetOrSequenceRecordsPerSec -> MeterWrapper(BrokerTopicStats.InvalidOffsetOrSequenceRecordsPerSec, "requests")
+    BrokerTopicStats.InvalidOffsetOrSequenceRecordsPerSec -> MeterWrapper(BrokerTopicStats.InvalidOffsetOrSequenceRecordsPerSec, "requests"),
```
Comment (satishd) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Why is `,` required here?

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 3]
```diff
@@ -386,7 +431,26 @@ class BrokerTopicMetrics(name: Option[String], configOpt: java.util.Optional[Kaf
...
 
-  def close(): Unit = metricTypeMap.values.foreach(_.close())
+  def close(): Unit = {
+    metricTypeMap.values.foreach(_.close())
+    metricGaugeTypeMap.values.foreach(_.close())
+  }
+}
+
+class BrokerTopicAggregatedMetric() {
+  private val partitionMetricValues = collection.concurrent.TrieMap[Int, Long]()
```
Comment (satishd) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Why is `TrieMap` used instead of  `javautil.concurrent.ConcurrentHashMap` here?

**Review by satishd — COMMENTED** [review]
Thanks @clolov for the PR, left a couple of minor comments.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 4]
```diff
@@ -283,8 +283,38 @@ class BrokerTopicMetrics(name: Option[String], configOpt: java.util.Optional[Kaf
...
+
+    def close(): Unit = gaugeLock synchronized {
+      if (lazyGauge != null) {
+        metricsGroup.removeMetric(metricType)
+        brokerTopicAggregatedMetric.close()
+        lazyGauge = null
+      }
+    }
+
+    gauge() // greedily initialize the general topic metrics
```
Comment (showuon) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: nit: I don't think we need comment here.

File: core/src/test/scala/integration/kafka/api/MetricsTest.scala [hunk 5]
```diff
@@ -320,16 +320,28 @@ class MetricsTest extends IntegrationTestHarness with SaslSetup {
...
+    s"kafka.server:type=BrokerTopicMetrics,name=$name"
+  }
+
   private def verifyRemoteStorageMetrics(shouldContainMetrics: Boolean): Unit = {
     val metrics = RemoteStorageMetrics.allMetrics().asScala.filter(name =>
       KafkaYammerMetrics.defaultRegistry.allMetrics.asScala.find(metric => {
         metric._1.getMBeanName().equals(name.getMBeanName)
       }).isDefined
     ).toList
+    val aggregatedBrokerTopicStats = Set(BrokerTopicStats.RemoteCopyLagBytes)
```
Comment (showuon) [inline core/src/test/scala/integration/kafka/api/MetricsTest.scala]: Why using set here? I guess we plan to add more metrics in this set in the future, is that right?

**Review by showuon — COMMENTED** [review]
@clolov , thanks for adding the integration tests. Left some comments. But I also hope we can have integration test that verify the `remoteCopyLag` metric will be updated with the expected value. You can refer to `RemoteLogManagerTest#testRemoteLogManagerTasksAvgIdlePercentMetrics` test to try to mock the size in localLogSegment and activeSegment to try to get some lag value there. Let me know if you have question.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 2 above)
Comment (clolov) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: It isn't 😬 - removed

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 3 above)
Comment (clolov) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: To be honest, I was looking for a concurrent implementation in Scala world and this is all I found - https://www.scala-lang.org/api/2.13.11/scala/collection/concurrent/index.html. I shall swap it for a concurrent map from Java world!

File: core/src/test/scala/integration/kafka/api/MetricsTest.scala (hunk 5 above)
Comment (clolov) [reply core/src/test/scala/integration/kafka/api/MetricsTest.scala]: Correct! I will add a few of the other metrics in the KIP via this aggregator - I just wanted to start with one to see whether people agree with it!

**Comment by clolov** [comment]
Heya @satishd and @showuon! I hope I have addressed your comments with the latest commits 😊

**Review by satishd — APPROVED** [review]
Thanks @clolov for addressing the review comments. LGTM.

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java [hunk 6]
```diff
@@ -699,22 +702,125 @@ void testRemoteLogManagerTasksAvgIdlePercentMetrics() throws Exception {
...
+
+        when(mockLog.onlyLocalLogSegmentsSize()).thenReturn(175L, 100L);
+        when(activeSegment.size()).thenReturn(100);
+
+        // before running tasks, the metric should not be registered
+        assertThrows(NoSuchElementException.class, () -> yammerMetricValue("RemoteCopyLogBytes"));
+        remoteLogManager.onLeadershipChange(Collections.singleton(mockLeaderPartition), Collections.singleton(mockFollowerPartition), topicIds);
+        TestUtils.waitForCondition(
+                () -> 75 == safeYammerMetricValue("RemoteCopyLagBytes"),
+                1000,
```
Comment (showuon) [inline core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java]: I think we don't need to change the wait time. Using default value makes it more reliable. We'll jump out the wait loop once the condition met anyway.

```
 TestUtils.waitForCondition(
                () -> 75 == safeYammerMetricValue("RemoteCopyLagBytes"),
                String.format("Expected to find 75 for RemoteCopyLagBytes metric value, but found %d", safeYammerMetricValue("RemoteCopyLagBytes")));
```

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java [hunk 7]
```diff
@@ -699,22 +702,125 @@ void testRemoteLogManagerTasksAvgIdlePercentMetrics() throws Exception {
...
+                String.format("Expected to find 75, but found %d", safeYammerMetricValue("RemoteCopyLagBytes")));
```
Comment (showuon) [inline core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java]: This error message is not clear. Maybe: 
`Expected to find 75 for RemoteCopyLagBytes metric value, but found %d"`?

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java [hunk 8]
```diff
@@ -699,22 +702,125 @@ void testRemoteLogManagerTasksAvgIdlePercentMetrics() throws Exception {
...
         return gauge.value();
     }
 
+    private long safeYammerMetricValue(String name) {
+        try {
+            return (long) yammerMetricValue(name);
+        } catch (NoSuchElementException ex) {
+            return 0L;
+        }
+    }
```
Comment (showuon) [inline core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java]: From the method name, we can't tell it's for long metric value. Maybe: `safeLongYammerMetricValue`?

**Review by showuon — COMMENTED** [review]
@clolov , thanks for the update. Overall LGTM! Left some minor comments.

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java [hunk 9]
```diff
@@ -699,22 +702,125 @@ void testRemoteLogManagerTasksAvgIdlePercentMetrics() throws Exception {
...
+        Map<MetricName, Metric> metricNameMetricMapAfter = KafkaYammerMetrics.defaultRegistry().allMetrics();
+        assertTrue((double) yammerMetricValue("RemoteLogManagerTasksAvgIdlePercent") < 1.0);
         // unlock copyLogSegmentData
         latch.countDown();
     }
 
-    private double yammerMetricValue(String name) {
-        Gauge<Double> gauge = (Gauge) KafkaYammerMetrics.defaultRegistry().allMetrics().entrySet().stream()
+    @Test
+    void testRemoteLogManagerRemoteCopyLagBytes() throws Exception {
```
Comment (showuon) [inline core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java:711]: Nice test. Thanks for adding it.

**Comment by clolov** [comment]
Heya @showuon! As always, your comments are spot on! I hope I have addressed them in the latest commits 😊

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the PR!

**Comment by kamalcph** [comment]
Will review the patch by today.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 1 above)
Comment (kamalcph) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: since we removed the `tags`. Do we need the method to initialise the gauge lazily? We are already instantiating the Gauge eagerly in L312.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 10]
```diff
@@ -319,11 +349,16 @@ class BrokerTopicMetrics(name: Option[String], configOpt: java.util.Optional[Kaf
         RemoteStorageMetrics.FAILED_REMOTE_FETCH_PER_SEC_METRIC.getName -> MeterWrapper(RemoteStorageMetrics.FAILED_REMOTE_FETCH_PER_SEC_METRIC.getName, "requests"),
         RemoteStorageMetrics.FAILED_REMOTE_COPY_PER_SEC_METRIC.getName -> MeterWrapper(RemoteStorageMetrics.FAILED_REMOTE_COPY_PER_SEC_METRIC.getName, "requests")
       ).asJava)
+      metricGaugeTypeMap.putAll(Map(
+        BrokerTopicStats.RemoteCopyLagBytes -> GaugeWrapper(BrokerTopicStats.RemoteCopyLagBytes, new BrokerTopicAggregatedMetric)
+      ).asJava)
```
Comment (kamalcph) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]:
```suggestion
        metricGaugeTypeMap.put(
            BrokerTopicStats.RemoteCopyLagBytes,
            GaugeWrapper(BrokerTopicStats.RemoteCopyLagBytes, new BrokerTopicAggregatedMetric))
```

Can we move the constant `RemoteCopyLagBytes` from `BrokerTopicStats` to `RemoteStorageMetrics`?

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 11]
```diff
@@ -368,6 +403,16 @@ class BrokerTopicMetrics(name: Option[String], configOpt: java.util.Optional[Kaf
...
 
+  def remoteCopyLagBytesWrapper: Gauge[Long] = metricGaugeTypeMap.get(BrokerTopicStats.RemoteCopyLagBytes).gauge()
+
+  def remoteBrokerTopicAggregateMetrics: Option[BrokerTopicAggregatedMetric] = {
+    if (metricGaugeTypeMap.contains(BrokerTopicStats.RemoteCopyLagBytes)) {
+      Option.apply(metricGaugeTypeMap.get(BrokerTopicStats.RemoteCopyLagBytes).brokerTopicAggregatedMetric)
+    } else {
+      Option.empty[BrokerTopicAggregatedMetric]
+    }
+  }
```
Comment (kamalcph) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala:407]: Can we provide a utility method to the caller?

```
  def recordRemoteCopyBytesLag(partition: Int, bytesLag: Long): Unit = {
    val brokerTopicAggregatedMetric = metricGaugeTypeMap.get(BrokerTopicStats.RemoteCopyLagBytes).brokerTopicAggregatedMetric
    brokerTopicAggregatedMetric.setPartitionMetricValue(partition, bytesLag)
  }
  ```
  and call:
  
  ```
  brokerTopicStats.topicStats(topic).recordRemoteCopyBytesLag(partition, lag);
  ```

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 12]
```diff
@@ -386,7 +431,26 @@ class BrokerTopicMetrics(name: Option[String], configOpt: java.util.Optional[Kaf
...
 
-  def close(): Unit = metricTypeMap.values.foreach(_.close())
+  def close(): Unit = {
+    metricTypeMap.values.foreach(_.close())
+    metricGaugeTypeMap.values.foreach(_.close())
+  }
+}
+
+class BrokerTopicAggregatedMetric() {
+  private val partitionMetricValues = new ConcurrentHashMap[Int, Long]().asScala
```
Comment (kamalcph) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]:
```suggestion
  private val partitionMetricValues = new ConcurrentHashMap[Int, Long]()
```

Can we keep the map in java itself?

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 13]
```diff
@@ -386,7 +431,26 @@ class BrokerTopicMetrics(name: Option[String], configOpt: java.util.Optional[Kaf
...
+
+  def setPartitionMetricValue(partition: Int, partitionValue: Long): Option[Long] = {
```
Comment (kamalcph) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Why do we have to return the value for `set` method?

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 14]
```diff
@@ -386,7 +431,26 @@ class BrokerTopicMetrics(name: Option[String], configOpt: java.util.Optional[Kaf
...
+    partitionMetricValues.put(partition, partitionValue)
+  }
+
+  def removePartition(partition: Int): Option[Long] = {
+    partitionMetricValues.remove(partition)
+  }
+
+  def value(): Long = partitionMetricValues.values.sum
```
Comment (kamalcph) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]:
```suggestion
  def value(): Long = partitionMetricValues.values().stream().mapToLong(v => v).sum
```

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 15]
```diff
@@ -404,6 +468,7 @@ object BrokerTopicStats {
   val ProduceMessageConversionsPerSec = "ProduceMessageConversionsPerSec"
   val ReassignmentBytesInPerSec = "ReassignmentBytesInPerSec"
   val ReassignmentBytesOutPerSec = "ReassignmentBytesOutPerSec"
+  val RemoteCopyLagBytes = "RemoteCopyLagBytes"
```
Comment (kamalcph) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Can this constant be moved to `RemoteStorageMetrics` similar to other metrics?

**Review by kamalcph — CHANGES_REQUESTED** [review]
Overall LGTM, left comments to address on code clarity. Will look into the tests once the comments are addressed. Thanks for working on this!

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 14 above)
Comment (kamalcph) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Also, instead of computing the aggregated `value` on each reporting (typically: 10 s). Can we keep one variable `value` and update it while recording the partition level metrics?

**Comment by showuon** [comment]
@clolov , could you address @kamalcph 's comments? I hope this KIP can be included into v3.7.0.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 13 above)
Comment (clolov) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: We do not - it isn't used currently!

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 12 above)
Comment (clolov) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Sure!

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 15 above)
Comment (clolov) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Oki

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 10 above)
Comment (clolov) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Oki

**Comment by clolov** [comment]
Heya, sorry, I will try to provide an updated version tomorrow

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 11 above)
Comment (clolov) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:407]: Sure, added a couple of utility methods!

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 14 above)
Comment (clolov) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: I actually have a preference for not updating one variable. The reason I don't want to use one variable is that a concurrent map provides fine-grained controls out of the box that no one else will be modifying the value while we read it. If I use a specific value then I have to acquire a lock. What are your thoughts?

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 1 above)
Comment (clolov) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Nope, it is no longer needed, I have since removed it 😊

**Comment by clolov** [comment]
Heya @kamalcph and @showuon - I hope I have addressed everything from the last review 😊

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java [hunk 16]
```diff
@@ -341,9 +341,10 @@ public void onLeadershipChange(Set<Partition> partitionsBecomeLeader,
 
             leaderPartitions.forEach(this::cacheTopicPartitionIds);
             followerPartitions.forEach(this::cacheTopicPartitionIds);
+            followerPartitions.forEach(
+                    topicIdPartition -> brokerTopicStats.topicStats(topicIdPartition.topic()).removeRemoteCopyBytesLag(topicIdPartition.partition()));
 
-            remoteLogMetadataManager.onPartitionLeadershipChanges(leaderPartitions, followerPartitions);
-            followerPartitions.forEach(topicIdPartition ->
+            remoteLogMetadataManager.onPartitionLeadershipChanges(leaderPartitions, followerPartitions);followerPartitions.forEach(topicIdPartition ->
```
Comment (kamalcph) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java]: please revert this change. Two statements in the same line

**Review by kamalcph — APPROVED** [review]
Patch and tests LGTM, thanks for addressing the review comments!

**Comment by showuon** [comment]
@clolov , there are checkstyle error, please correct them:
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14832/6/pipeline

Thanks.

**Comment by showuon** [comment]
@clolov , I took the liberty to help fix the checksytle error. Let's see if the CI build passed. Thanks.

**Comment by showuon** [comment]
It failed 61 tests in the CI build. Re-triggering it:
https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14832/8/

**Comment by showuon** [comment]
Failed tests are unrelated:
```
    Build / JDK 21 and Scala 2.13 / org.apache.kafka.connect.integration.OffsetsApiIntegrationTest.testAlterSinkConnectorOffsets
    Build / JDK 21 and Scala 2.13 / kafka.api.TransactionsBounceTest.testWithGroupMetadata()
    Build / JDK 21 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testFenceMultipleBrokers()
    Build / JDK 8 and Scala 2.12 / kafka.api.ConsumerBounceTest.testConsumptionWithBrokerFailures()
    Build / JDK 8 and Scala 2.12 / kafka.api.TransactionsBounceTest.testWithGroupMetadata()
    Build / JDK 8 and Scala 2.12 / kafka.server.DynamicBrokerReconfigurationTest.testTrustStoreAlter(String).quorum=kraft
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManagerTest.testNewPartitionUpdates()
    Build / JDK 11 and Scala 2.13 / kafka.server.DelegationTokenRequestsTest.testDelegationTokenRequests(String).quorum=kraft
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.testDescribeQuorumStatusSuccessful [2] Type=Raft-Isolated, MetadataVersion=3.7-IV3, Security=PLAINTEXT
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.common.security.authenticator.SaslAuthenticatorFailureNoDelayTest.testInvalidPasswordSaslScram()
    Build / JDK 17 and Scala 2.13 / kafka.api.PlaintextAdminIntegrationTest.testElectPreferredLeaders(String).quorum=kraft
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.EosV2UpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosV2[true]
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.testDescribeQuorumReplicationSuccessful [2] Type=Raft-Isolated, MetadataVersion=3.7-IV3, Security=PLAINTEXT
```
